Misc Convenience Helpers

contagon opened the pull request:
This is a handful of feature I found helpful in my own code using JRL. Figured I'd share if you think it's worthwhile having in the main codebase. These include:

- `addGroundTruth` and `addInitialization` functions in DatasetBuilder. This allows for setting these independently since they're not saved in the `Entry` anyways.
- Custom parsers/serializers support. There was a stub that needed filling out for this and it just took a couple lines of code. Also added tests for it.
- filtering/removing certain measurement types from `Entry`. For my purposes, I added all my wheel factors to a single jrl file, then used the filter functionality to only run on the one I was interested in. Tests also included.

Let me know what you think! Happy to split this up however if you're interested in some features, but not others.

DanMcGann reviewed (changes requested):
Most looks good!

There are a few small things (see comments below) that should be addressed.

Additionally, Id like to discuss is the filter interface. I see this as being very useful!! but I think it could be improved from its current implementation. 

Firstly, I find the current naming a little confusing. The difference between "filter" / "remove" was unclear until I read the docs and having separate functions for the positive predicate and negative predicate feels off to me since they are both really just "filters"

My proposal is to have a single `filtered` function (past tense to imply it is returning a filtered version and not modifying the object in place). `filtered` accepts a predicate function that is used to remove out measurements. Then we can have some default predicate functors `keep_types` and `remove_types` to use with `filtered` to realize the functionality that is proposed in this PR.

I also see this as being really useful if exposed in the `Dataset` class. However, filtering arbitrary measurements may affect the variables observed in a dataset, and figuring which variables to remove would be pretty tricky :/

contagon commented on tests/test_Custom.cpp:
```diff
@@ -0,0 +1,78 @@
+    parser.registerMeasurementParser(customType, [](const json& input){ return jrl::io_measurements::parsePrior<gtsam::Pose2>(&jrl::io_values::parse<gtsam::Pose2>, input); });
+
+    // Send it round trip!
+    writer.writeDataset(builder.build(), "custom.jrl");
```
It gets written into the 'build/tests' directory. Currently does not get removed, but that'd be an easy fix.

I'll plan on exposing `serializeMeasurements` and `parseMeasurements` and test via those instead.

I would argue that it'd be worthwhile to have some tests write to file, otherwise the functions `writeDataset`, `writeJson`, etc won't ever be tested which is the interface most users will be using. That probably belongs in another test though!

contagon commented:
Using a predicate function is a great idea - I'll definitely rewrite it in that form, shouldn't be too hard. I'll tweak the tests as I do that as well.

contagon commented:
K, just made those changes, give it another look when you get the chance,
- Updated parse and serialize measurement/values functions to be public, updated test to use those instead of file
- Made `filtered` function, along with `KeepTypes` and `RemoveTypes` static methods in `Entry`
- Ran clang-format

DanMcGann commented on include/jrl/Types.h:
```diff
@@ -1,9 +1,11 @@
 namespace jrl {
 
 typedef std::map<gtsam::Key, std::string> ValueTypes;
+typedef std::function<bool(const std::string&, const gtsam::NonlinearFactor::shared_ptr&)> EntryPredicate;
```
Since this is only used in the context of an entry lets move it within that class can rename it to FilterPredicate.
Externally it can be accessed with `Entry::FilterPredicate`.

DanMcGann reviewed (changes requested):
Changes look great!
Just one little thing on where to define the predicate type that I think would be best to change before merging

contagon commented:
Made that tweak and merged from main, I think it should be good now.

DanMcGann reviewed (approved):
all looks good!
Thanks for dealing with all my change requests!
